Bluetooth: GATT: Rework Service Changed indications

Vudentz opened the pull request:
There could be situations where many services are changed in a row which
would cause k_sem_take to block on the second change, but if the calling
thread is actually the RX thread then this will deadlock since the RX
thread is the one processing the confirmations of indications and it is
blocked k_sem_give is never called.

To solve this the services changes are now offloaded to the system wq
and the code will attempt to consolidate the range being changed so only
one indication is send. If for some reason another changes is caused
while confirmation is pending we just reschedule it to run later to avoid
blocking the system wq in the same way.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Rework Service Changed indications

There could be situations where many services are changed in a row which
would cause k_sem_take to block on the second change, but if the calling
thread is actually the RX thread then this will deadlock since the RX
thread is the one processing the confirmations of indications and it is
blocked k_sem_give is never called.

To solve this the services changes are now offloaded to the system wq
and the code will attempt to consolidate the range being changed so only
one indication is send. If for some reason another changes is caused
while confirmation is pending we just reschedule it to run later to
avoid blocking the system wq in the same way.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -31,6 +31,8 @@
 #include "smp.h"
 #include "gatt_internal.h"
 
+#define SC_TIMEOUT K_MSEC(200)
```
I don't really like having to add arbitrary timeouts to the flow of events

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -128,46 +130,95 @@ static int gatt_register(struct bt_gatt_service *svc)
+	struct bt_gatt_indicate_params params;
+	u16_t start;
+	u16_t end;
+	struct k_sem sem;
```
The semaphore seems a bit overkill. At most you'll need a flag to say that the procedure should be redone once the confirmation to a pending indication is received.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -128,46 +130,95 @@ static int gatt_register(struct bt_gatt_service *svc)
 	BT_DBG("err 0x%02x", err);
 
-	k_sem_give(&sc_sem);
+	k_sem_give(&gatt_sc.sem);
```
Instead of giving a semaphore here, I think it'd make sense to have a flag "services changed after last SC indication was sent" and if the flag is set resend another indication from the response callback.

Vudentz commented:
New revision with the following changes:

- SC_TIMEOUT is now just 10 msec since we just works as a idle timeout it should be as low as possible.
- Replace k_sem with an atomic flag as that is cheaper and we don't really more than that.

Vudentz committed:
Bluetooth: Shell: Implement support for gatt-metrics off

With introduction of bt_gatt_service_unregister it is now possible to
unregister service at runtime.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: Shell: Add second vendor service

This adds a second vendor service testing if service changed indications
works with more than one change in a row and as a bonus it implements
echo attribute which notifies any data that is written to it:

> ACL Data TX: Handle 3585 flags 0x00 dlen 9
      ATT: Write Command (0x52) len 4
        Handle: 0x0013
          Data: 0000
< ACL Data RX: Handle 3585 flags 0x02 dlen 9
      ATT: Handle Value Notification (0x1b) len 4
        Handle: 0x0013
          Data: 0000

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
Added a few other changes:

- Introduce a second flag so we can avoid resubmits and just wait until indicate is complete.
- Add a few commands to the shell testing the correct behavior of multiple service changes in a row.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -128,46 +130,108 @@ static int gatt_register(struct bt_gatt_service *svc)
+	/* Check if there is new change in the meantime */
+	if (atomic_test_bit(gatt_sc.flags, SC_RANGE_CHANGED)) {
+		/* Reschedule without any delay since it is waiting already */
+		k_delayed_work_submit(&gatt_sc.work, 0);
```
K_NO_WAIT instead of 0

Vudentz replied:
Done

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -128,46 +130,108 @@ static int gatt_register(struct bt_gatt_service *svc)
-		BT_DBG("Service Changed indicating, waiting until complete...");
-		k_sem_take(&sc_sem, K_FOREVER);
+	if (atomic_test_bit(sc->flags, SC_INDICATE_PENDING)) {
+		BT_DBG("indicate pending, waiting until complete...");
```
Shouldn't this be an assert?

Vudentz replied:
Done
